src/formatter.rs

```diff
@@ -2,77 +2,111 @@
 
 use std::io::BufRead;
 
-use ansi_term::{ANSIString, ANSIStrings};
 use log::debug;
 
-use crate::config::Config;
+use crate::config::StyleConfig;
+use crate::extensions::FindFrom;
 use crate::tokenizer::Tokenizer;
 use crate::types::LineType;
 
-fn highlight_command<'a>(
-    command: &'a str,
-    example_code: &'a str,
-    config: &Config,
-    parts: &mut Vec<ANSIString<'a>>,
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum HighlightingSnippet<'a> {
+    CommandName(&'a str),
+    Variable(&'a str),
+    NormalCode(&'a str),
+    Description(&'a str),
+    Text(&'a str),
+    Linebreak,
+}
+
+impl<'a> HighlightingSnippet<'a> {
+    pub fn is_empty(&self) -> bool {
+        use HighlightingSnippet::*;
+
+        match self {
+            CommandName(s) | Variable(s) | NormalCode(s) | Description(s) | Text(s) => s.is_empty(),
+            Linebreak => false,
+        }
+    }
+}
+
+/// Checks whether the characters right before and after the substring (given by half-open index interval) are whitespace (if they exist).
+fn is_freestanding_substring<'a>(surrouding: &'a str, substring: (usize, usize)) -> bool {
+    let (start, end) = substring;
+    surrouding
+        .char_indices()
+        .take_while(|&(i, _)| i < start)
+        .last()
+        .filter(|(_, prev_char)| !prev_char.is_whitespace())
+        .is_none()
+        && surrouding[end..]
+            .chars()
+            .next()
+            .filter(|next_char| !next_char.is_whitespace())
+            .is_none()
```

Comment:
For more clarity maybe assign both conditions to (named) variables, e.g. `char_before_start_is_whitespace` and `char_after_end_is_whitespace`.

Also, the doubled negation can be avoided with `....filter(|c| c.is_whitespace()).is_some()`.

Reply:
Actually, if there is no char before the substring, this would be `None` too, which is why I wrote it like this. It's the best way I found for actually doing the "doesn't exist or is whitespace" check in one go.

I added the names, maybe it's clearer with that?

Reply:
I got your point with the `None` case. The variables make things clearer, indeed. As both checks do the same thing we could also introduce another method implementing what the comment explains to get rid of it (some people don't like comments :smile:), like:
```rust
fn is_whitespace_or_does_not_exist(char: Option<char>) -> bool {
    char.filter(|c| !c.is_whitespace()).is_none()
}

fn is_freestanding_substring(surrouding: &str, substring: (usize, usize)) -> bool {
    let (start, end) = substring;
    let char_before = surrouding[..start].chars().last();
    let char_after = surrouding[end..].chars().next();
    is_whitespace_or_does_not_exist(char_before) && is_whitespace_or_does_not_exist(char_after)
}
```

What do you think?

Reply:
hm, I don't know, seems kind of extreme to me :man_shrugging: I would find this a bit weird as its own function

Reply:
Fine for me.